pandas/tests/reshape/test_concat.py

```diff
@@ -829,12 +831,128 @@ def test_append_preserve_index_name(self):
+        result = df.append(ser)
+        expected = pd.DataFrame([[1., 2., 3.], [4, 5, 6], [7, 8, np.nan]],
+                                index=[0, 1, 2],
+                                columns=df_columns)
+        assert_frame_equal(result, expected)
+
+        # ser wider than df
+        ser_index = df_columns
+        df_columns = df_columns[:2]
+        df = pd.DataFrame([[1, 2], [4, 5]], columns=df_columns)
+        ser = pd.Series([7, 8, 9], index=ser_index, name=2)
+        result = df.append(ser)
+        expected = pd.DataFrame([[1, 2, np.nan], [4, 5, np.nan], [7, 8, 9]],
+                                index=[0, 1, 2],
+                                columns=ser_index)
+        assert_frame_equal(result, expected)
+
+    @pytest.mark.parametrize("df_columns, series_index", combinations([
+        pd.RangeIndex(3),
+        pd.Index([4, 5, 6]),
+        pd.Index([7.5, 8.5, 9.5]),
+        pd.CategoricalIndex('A B C'.split()),
+        pd.DatetimeIndex([dt.datetime(2013, 1, 3, 0, 0),
+                          dt.datetime(2013, 1, 3, 6, 10),
+                          dt.datetime(2013, 1, 3, 7, 12)]),
+    ], r=2), ids=lambda x: str(x.dtype))
+    def test_append_different_columns_types(self, df_columns, series_index):
+        # GH18359
+        # see also tests 'test_append_multi_index_raises' and
+        # 'test_append_interval_index_raises' below
+
+        df = pd.DataFrame([[1, 2, 3], [4, 5, 6]], columns=df_columns)
+        ser = pd.Series([7, 8, 9], index=series_index, name=2)
+
+        result = df.append(ser)
+        idx_diff = ser.index.difference(df_columns)
+        combined_columns = df_columns.astype(object).append(idx_diff)
+        expected = pd.DataFrame([[1., 2., 3., np.nan, np.nan, np.nan],
+                                 [4, 5, 6, np.nan, np.nan, np.nan],
+                                 [np.nan, np.nan, np.nan, 7, 8, 9]],
+                                index=[0, 1, 2],
+                                columns=combined_columns)
+        assert_frame_equal(result, expected)
+
+    @pytest.mark.parametrize("other_type", [
+        pd.RangeIndex(3),
+        pd.CategoricalIndex('A B C'.split()),
+        pd.CategoricalIndex('A B C'.split(), ordered=True),
+        pd.IntervalIndex.from_breaks([0, 1, 2, 3]),
+        pd.DatetimeIndex([dt.datetime(2013, 1, 3, 0, 0),
+                          dt.datetime(2013, 1, 3, 6, 10),
+                          dt.datetime(2013, 1, 3, 7, 12)]),
+        pd.Index([4, 5, 6]),
+    ], ids=lambda x: str(x.dtype))
+    def test_append_multi_index_raises(self, other_type):
+        # GH18359
+        # .append will raise if MultiIndex appends or is appended to a
+        # different index type
+
+        mi = pd.MultiIndex.from_arrays(['A B C'.split(), 'D E F'.split()])
+
+        df = pd.DataFrame([[1, 2, 3], [4, 5, 6]], columns=mi)
+        ser = pd.Series([7, 8, 9], index=other_type, name=2)
+        if isinstance(other_type, pd.IntervalIndex):
```

Comment:
so this is wrong, append should raise a TypeError (like all the others). can you fix this in II?

Reply:
It looks like the `ValueError` is coming from here:

https://github.com/pandas-dev/pandas/blob/1245f0673074316b53d1d1ee10abd6077058b133/pandas/core/indexes/interval.py#L1155-L1161

Seems reasonable to split the `if` condition, and raise a different exception in each case, along the lines of:
```python
if not isinstance(other, IntervalIndex):
    raise TypeError(error_msg)
elif self.closed != other.closed:
    raise ValueError(error_msg)
```

I think this change will only break the test below, at least within the II tests, but all that needs to be changed is `ValueError` --> `TypeError`:
https://github.com/pandas-dev/pandas/blob/1245f0673074316b53d1d1ee10abd6077058b133/pandas/tests/indexes/interval/test_interval.py#L936-L940